core/lib/state/src/cache/sequential_cache.rs

```diff
@@ -0,0 +1,150 @@
+use std::collections::VecDeque;
+
+use crate::cache::metrics::{Method, RequestOutcome, METRICS};
+
+/// A generic cache structure for storing key-value pairs in sequential order.
+/// It allows for non-unique keys and supports efficient retrieval of values based on a key
+/// threshold. The cache maintains a specified maximum capacity, removing the oldest entries
+/// as new ones are added.
+///
+/// Usage example: storing mempool transactions and querying them by the received_at field
+/// (clients maintain a cursor based on this key and poll the data structure for newer transactions)
+///
+///
+/// K: Type of the ordered, potentially non-unique key. Example: Transaction's received_at field.
+/// V: Type of the value associated with each key. Example: Transaction's hash.
+#[derive(Debug, Clone)]
+pub struct SequentialCache<K, V> {
+    name: &'static str,
+    data: VecDeque<(K, V)>,
+    capacity: usize,
+}
+
+impl<K: Ord + Copy, V: Clone> SequentialCache<K, V> {
+    /// Creates a new `SequentialCache` with the specified maximum capacity.
+    pub fn new(name: &'static str, capacity: usize) -> Self {
+        assert!(capacity > 0, "Cache capacity must be greater than 0");
+        SequentialCache {
+            name,
+            data: VecDeque::with_capacity(capacity),
+            capacity,
+        }
+    }
+
+    /// Inserts multiple key-value pairs into the cache from an iterator. If adding these
+    /// items exceeds the cache's capacity, the oldest entries are removed. Keys can be non-unique.
+    /// Panics when keys order is incorrect (a smaller key is inserted after a larger one)
+    pub(crate) fn insert<I>(&mut self, items: I)
+    where
+        I: IntoIterator<Item = (K, V)>,
+    {
+        for (key, value) in items {
+            let latency = METRICS.latency[&(self.name, Method::Insert)].start();
+            if let Some(last_key) = self.get_last_key() {
+                assert!(key >= last_key, "Keys must be inserted in sequential order");
+            }
```

Comment:
Nit: Given that we're (generally) try to avoid panicking right now, and given that it's not an internal invariant, probably it makes sense for this function to return `Result`. It's external code to guarantee the order of keys, so `cache.insert(items).expect("DAL guarantees keys to be ordered");` in theory makes more sense

Reply:
Also `Option` implements `Ord` for `T: Ord`, so it can be a one-liner, e.g. `anyhow::ensure!(Some(key) >= self.get_last_key(), "Keys must be inserted in sequential order");`

Reply:
> Option implements Ord for T: Ord

wow, really didn't expect it. Is surprising for me since `None < Some(1)` or `None > Some(1)` is essentially arbitrary (one can argue though that None can be thought of as `0`, hence it's less than anything. 

Found [this](https://www.reddit.com/r/rust/comments/1bdqocw/why_is_ord_implemented_on_option/) thread with many explanations, so I'm good :)

Will change to `result` as well.